commands/command_migrate_export.go

```diff
@@ -0,0 +1,155 @@
+			if filepath.Base(path) == ".gitattributes" {
+				return b, nil
+			}
+
+			var buf bytes.Buffer
+
+			if _, err := smudge(gitfilter, &buf, b.Contents, path, false, rewriter.Filter()); err != nil {
+				return nil, err
+			}
+
+			return &odb.Blob{
+				Contents: &buf, Size: int64(buf.Len()),
+			}, nil
+		},
+
+		TreeCallbackFn: func(path string, t *odb.Tree) (*odb.Tree, error) {
+			if path != "/" {
+				// Ignore non-root trees.
+				return t, nil
+			}
+
+			ours := tracked
+			theirs, err := trackedFromAttrs(db, t)
+			if err != nil {
+				return nil, err
+			}
+
+			// Create a blob of the attributes that are optionally
+			// present in the "t" tree's .gitattributes blob, and
+			// union in the patterns that we've tracked.
+			//
+			// Perform this Union() operation each time we visit a
+			// root tree such that if the underlying .gitattributes
+			// is present and has a diff between commits in the
+			// range of commits to migrate, those changes are
+			// preserved.
+			blob, err := trackedToBlob(db, theirs.Clone().Union(ours))
+			if err != nil {
+				return nil, err
+			}
+
+			// Finally, return a copy of the tree "t" that has the
+			// new .gitattributes file included/replaced.
+			return t.Merge(&odb.TreeEntry{
+				Name:     ".gitattributes",
+				Filemode: 0100644,
+				Oid:      blob,
+			}), nil
+		},
+
+		UpdateRefs: true,
+	}
+
+	requireInRepo()
+
+	opts, err = rewriteOptions(args, opts, l)
+	if err != nil {
+		ExitWithError(err)
+	}
+
+	// If we have a valid remote, pre-download all objects using the Transfer Queue
+	if remoteURL := getAPIClient().Endpoints.RemoteEndpoint("download", cfg.Remote()).Url; remoteURL != "" {
+		q := newDownloadQueue(getTransferManifestOperationRemote("Download", cfg.Remote()), cfg.Remote())
+		if err := rewriter.ScanForPointers(q, opts, gitfilter); err != nil {
+			ExitWithError(err)
+		}
+
+		q.Wait()
+
+		for _, err := range q.Errors() {
+			if err != nil {
+				ExitWithError(err)
+			}
+		}
+	}
+
+	// Perform the rewrite
+	if _, err := rewriter.Rewrite(opts); err != nil {
+		ExitWithError(err)
+	}
+
+	// Only perform `git-checkout(1) -f` if the repository is
+	// non-bare.
+	if bare, _ := git.IsBare(); !bare {
+		t := l.Waiter("migrate: checkout")
+		err := git.Checkout("", nil, true)
+		t.Complete()
+
+		if err != nil {
+			ExitWithError(err)
+		}
+	}
+}
+
+// trackedFromExportFilter returns an ordered set of strings where each entry is a
+// line in the .gitattributes file. It adds/removes the fiter/diff/merge=lfs
+// attributes based on patterns included/excluded in the given filter. Since
+// `migrate export` removes files from Git LFS, it will remove attributes for included
+// files, and add attributes for excluded files
+func trackedFromExportFilter(filter *filepathfilter.Filter) *tools.OrderedSet {
+	tracked := tools.NewOrderedSet()
+
+	for _, include := range filter.Include() {
+		tracked.Add(fmt.Sprintf("%s text -filter -merge -diff", escapeAttrPattern(include)))
```

Comment:
Nice. I think that we are supposed to spell the `-filter` as `!filter`, though.

Reply:
According to the [gitattributes documentation](https://git-scm.com/docs/gitattributes), it looks like `-` sets the value of an attribute to `false`, whereas `!` sets the value of an attribute to `unspecified`. Since we're doing an export here, it seems like it might be better to use `-`

Reply:
That's what it looks like to me, too:

https://github.com/git/git/blob/53f9a3e157dbbc901a02ac2c73346d375e24978c/attr.c#L334-L335

But, I think that `ATTR__UNSET` is more appropriate; according to the merge documentation from that same page:

> ### Unset
> Take the version from the current branch as the tentative merge result, and
> declare that the merge has conflicts. This is suitable for binary files that
> do not have a well-defined merge semantics.
>
> ### Unspecified
> By default, this uses the same built-in 3-way merge driver as is the case when
> the `merge` attribute is set. However, the `merge.default` configuration
> variable can name different merge driver to be used with paths for which the
> `merge` attribute is unspecified.

I think that given the choice between the two, we'd prefer unspecified, which defaults back to the configuration.

Reply:
Ah, that makes sense! In that case I'll go ahead and switch this over